templates/skeleton/TEMPLATE_GUIDELINES.md

```diff
@@ -0,0 +1,251 @@
+
+## Error Handling
+
+Always demonstrate realistic error-handling. Skeleton templates should be a shining example of handling any sort of error that can occur. This will also help teach developers what kinds of errors could occur, where they occur, and how to handle them correctly.
+
+### Do:
+
+- **Have an `ErrorBoundary` in every route template.** `ErrorBoundary` is used when an Error is thrown in a “loader”, and is generally meant for unexpected errors, like 500, 503, etc. Any Storefront query or mutation error will be handled by the `ErrorBoundary`. Type the error as “unknown” since _anything_ in JS can be thrown 🙂
+- **Use the “errorElement” prop on every `<Await>` component.** When using “defer”, some promises may be rejected at a later time. The only way to handle this is to use the “errorElement” on the associated <Await> component, otherwise the error is swallowed.
+- **Use try/catch** – except in “loader”, “action”, and the Component. Those three “Route Module APIs” are handled automatically by `ErrorBoundary` and CatchBoundary, but the rest – such as “meta”, “links”, “handle”, etc. – will crash the server if an error is thrown.
+- **Have a CatchBoundary if necessary.** A CatchBoundary is used when a new Response is thrown, and is generally meant for expected errors caused by the user, such as 401, 404, etc. Note that `CatchBoundary`s will be deprecated in Remix V2, at which time we'll remove this recommendation.
+
+### Don’t:
+
+- **Use try/catch in “loader”, “action”, and the Component.** For templates it’s easier to let the error be thrown and get handled by the `ErrorBoundary` than to handle it manually.
+
+### Examples
+
+PR: https://github.com/Shopify/hydrogen/pull/716
+
+Code:
+
+```tsx
+export async function loader() {
+  /* failed SFAPI requests will throw to the ErrorBoundary */
+  const data = storefront.query()
+
+
+  /* If no page data is returned purposely throw to the CatchBoundary */
+  if (!data?.page) {
+    throw new Response('Page not found', { status: 404 })
+  }
+
+  //...
+  return defer()
+}
+
+export function meta() {
+  try {
+    //
+  catch(error) {
+    //
+  }
+}
+
+export function ErrorBoundary({error}) {
+  return (<div>{error.message}</div>)
+}
+
+export function CatchBoundary() {
+  const {statusText} = useCatch()
+  return (<div>{statusText}</div>)
+}
+
+export default function TheUIComponents() {
+  return (
+    <Await resolve={} errorElement={<div>An error occurred</div>}>
+      //
+    </Await>
+  )
+}
+```
+
+## Template Dependencies
+
+Minimalize dependencies and shared code between templates.
+
+### Do:
+
+- Keep all code associated with a template in a single template file, even if that means some duplication in templates
+
+### Don't:
+
+- Share code between templates in a separate file. There is no guarantee that any given project will need both those templates, and we also want to avoid issues associated with generating multiple files (such as file placement)
```

Comment:
Yeah that came out bad. What I meant was having a helper file that's generated somewhere under `routes/` so that it ends up becoming a Remix route by accident.

Reply:
```suggestion
- Share code between templates in a separate file. There is no guarantee that any given project will need both those templates, and we also want to avoid issues associated with generating multiple files (such as file organization).
```